### PR TITLE
GUI: Peer table: Visualize inbound/outbound state for every row

### LINKED ISSUE #13483: Remote peer debug window: Show connection direction (in/out)
## Preamble
Bitcoin-qt 0.16.1

In the GUI menue under
Help->Debug window->Remote Peers
the remote peers are listed.

In the "Node/Service" column, the remote peer IP or .onion address and TCP port are listed.

## Current situation
Currently, the information about the direction of the connection to/from that remote peer can only be seen when clicking on the according peer row in the table. Then, on the right is displayed "Direction: incoming" (or "outgoing").

## Suggestion for enhancement
To see the direction quick and easily for all remote peers at once, and so have a quick comfortable overview on the connections situation, I would suggest the following enhancement:

Left to the remote peer IP or .onion address (in the same "Node/Service" column),
display an upwards arrow ↑ for outgoing connections,
or a downwards arrow ↓ for incoming connections.
E.g. `↑ 10.20.30.40:8333` for outgoing connection to IP 10.20.30.40 on TCP port 8333,
or `↓ 10.20.30.40:13149` for incoming connection from 10.20.30.40:13149

or alternatively
Add a new column, which shows the direction (either graphical symbols like arrows, or words like "incoming" or "outgoing")

### PR DESCRIPTION
**Pull request by ghost**
Fixes #13483

The address in the network peer table is prefixed with an up-arrow symbolizing an outbound connection, or an down-array symbolizing an inbound connection. See screenshot.

The user has an easy visual confirmation about the connection direction state. I really like it :)
Impact to columns sorting is grouping by inbound/outbound first, which in my opinion is an advantage, too.
bildschirmfoto

### CODE REVIEW

File: src/qt/peertablemodel.cpp [hunk 1]
```diff
@@ -157,12 +157,17 @@ QVariant PeerTableModel::data(const QModelIndex &index, int role) const
...
+        // peer address output prefix will be an arrow symbol of inbound/outbound
+        QString address_output_prefix;
+
         switch(index.column())
         {
         case NetNodeId:
             return (qint64)rec->nodeStats.nodeid;
         case Address:
-            return QString::fromStdString(rec->nodeStats.addrName);
+            rec->nodeStats.fInbound ? address_output_prefix = "↓ " : address_output_prefix = "↑ ";
```
Comment (promag) [inline src/qt/peertablemodel.cpp]: Prefer:
```cpp
address_output_prefix = rec->nodeStats.fInbound ? "↓ " : "↑ ";
```

File: src/qt/peertablemodel.cpp [hunk 2]
```diff
@@ -157,12 +157,17 @@ QVariant PeerTableModel::data(const QModelIndex &index, int role) const
...
+            return QString::fromStdString(rec->nodeStats.addrName).prepend(address_output_prefix);
```
Comment (promag) [inline src/qt/peertablemodel.cpp]: Could inline like:
```cpp
return QString(rec->nodeStats.fInbound ? "↓ " : "↑ ") + QString::fromStdString(rec->nodeStats.addrName);
```
and remove declaration above.

**Review by promag — COMMENTED** [review]
Concept ACK.

**Comment by wodry** [comment]
@promag Nice code shortening. Rebased and squashed.

**Comment by wodry** [comment]
What about the following idea which had the advantage of easier visual recognition, because one sees the difference at the first character ("arrow or not ?" in opposite to "up arrow or down arrow?")?

inbound:
`→ 1.2.3.4:8333`

outbound:
`1.2.3.4:45376 ←`

File: src/qt/peertablemodel.cpp [hunk 3]
```diff
@@ -162,7 +162,8 @@ QVariant PeerTableModel::data(const QModelIndex &index, int role) const
         case NetNodeId:
             return (qint64)rec->nodeStats.nodeid;
         case Address:
-            return QString::fromStdString(rec->nodeStats.addrName);
+            // prepend to peer address down-arrow symbol for inbound connection and up-arrow for outbound connection
+            return QString(rec->nodeStats.fInbound ? "↓ " : "↑ ") + QString::fromStdString(rec->nodeStats.addrName);
```
Comment (jonasschnelli) [inline src/qt/peertablemodel.cpp]: Does those unicode char (86 93 and 86 91) work on all supported platforms (especially older windows version)?
Would using an Icon make more sense?

**Review by jb55 — APPROVED** [review]
utACK 4132ad3bfa5b5ead3f4a592d819d0097881b4827

File: src/qt/peertablemodel.cpp (hunk 3 above)
Comment (wodry) [reply src/qt/peertablemodel.cpp]: Could You define 'older'?

Comment (laanwj) [reply src/qt/peertablemodel.cpp]: These characters have been in unicode since time immemorial, I wouldn't be afraid of them not being supported.

**Comment by jonasschnelli** [comment]
utACK 4132ad3bfa5b5ead3f4a592d819d0097881b4827

**Comment by laanwj** [comment]
> The address in the network peer table is prefixed with an up-arrow symbolizing an outbound connection, or an down-array symbolizing an inbound connection. See screenshot.

Thank you for doing it this way, instead of adding another column.
utACK 4132ad3bfa5b5ead3f4a592d819d0097881b4827

**Comment by fanquake** [comment]
tACK 4132ad3

#### macOS 10.13.5
master:

macos-master

4132ad3:

macos-13537

#### Windows 10
4132ad3:

windows10-13537

The symbol could be a bit hard to make out on Windows, but Windows really needs an overhaul in terms of font/sizing anyways, so I don't think that should block anything here.

**Comment by hebasto** [comment]
> Impact to columns sorting is grouping by inbound/outbound first, which in my opinion is an advantage, too.

It is removes an option to sort by address, that is wrong.

**Pull request merged**